src/core/maps/grid_map.h

```diff
@@ -68,18 +87,126 @@ class GridMap {
+                (cell_coord.x + _map_center_x) < _width &&
+           0 <= (cell_coord.y + _map_center_y)          &&
+                (cell_coord.y + _map_center_y) < _height;
+  }
+
+
+  void update_size(const DiscretePoint2D& cell_coord) {
+    resize_bound(RESIZE_HORZ, cell_coord.x+_map_center_x);
+    resize_bound(RESIZE_VERT, cell_coord.y+_map_center_y);
+  }
+
+  void resize_bound(const int res_dim, const int container_coord) {
+    if (res_dim != RESIZE_HORZ && res_dim != RESIZE_VERT)
+      return;
+    int bound = (res_dim == RESIZE_HORZ ? width() : height());
+    //find if it needs to resize map
+    int new_bound = calc_sufficient_bound(container_coord, bound);
+    if (new_bound == bound)
+      return;
+    //expands map with the exponential rule
+    // states are +100%, +300%, +700% etc ( -100% + [2^n*100%] )
+    int expand_coef = closest_bounded_power_two(new_bound/bound) - 1;
+    resize_in_direction(expand_coef*bound,
+                        res_dim<<RESIZE_DIM_BIT | (container_coord<0) );
+  }
+
+  int calc_sufficient_bound(const int container_coord, const int map_bound) {
+    if (container_coord < 0)
+      return (-container_coord)+map_bound;
+    if (map_bound <= container_coord)
+      return container_coord + 1;
+    return map_bound;
+  }
+
+  /* Finds the nearest upper bound value of degree of 2 to the input integer
+   * NOTE: 2-3  are bounded by 4,
+   *       4-7  are bounded by 8,
+   *       8-15 are bounded by 16 etc.
+   */
+  long closest_bounded_power_two(long x) {
+    long p2 = 1;
+    while (p2 <= x) {
+      p2 <<= 1;
+    }
+    return p2;
+  }
+
+  void resize_in_direction(const int delta, const int direction_flag) {
+
+    if (delta <= 0)
+      return;
+
+    if (direction_flag == RESIZE_DOWN) {
+      add_empty_rows(delta, _cells.begin());
+      _map_center_y  += delta;
+    } else if (direction_flag == RESIZE_UP) {
+      add_empty_rows(delta, _cells.end());
+    } else if (direction_flag == RESIZE_LEFT) {
+      add_empty_cols(delta, [](Row& vector) { return vector.begin(); } );
+      _map_center_x += delta;
+    } else if (direction_flag == RESIZE_RIGHT) {
+      add_empty_cols(delta, [](Row& vector) { return vector.end(); } );
+    }
+  }
+
+  void add_empty_rows(const int expand_value, Map::iterator pos) {
+    Map empty_rows(expand_value, Row(width()));
+    for (auto& row : empty_rows) {
+      for (auto& cell : row) {
+        cell = _cell_factory->create_cell();
+      }
+    }
+
+    _cells.insert(pos, empty_rows.begin(), empty_rows.end());
+    _height += expand_value;
+  }
+
+  void add_empty_cols(const int expand_value,
+                      std::function<Row::iterator(Row&)> pos) {
+    Map empty_cols(height(), Row(expand_value));
+    for (auto& row : empty_cols) {
+      for (auto& cell : row) {
+        cell = _cell_factory->create_cell();
+      }
+    }
+
+    for (size_t i = 0; i < empty_cols.size(); i++) {
+      _cells[i].insert(pos(_cells[i]),
```

Comment:
try to write the statement on a single line. You may try to move empty_cell[i] to a variable with a shorter name.